File: strings/28. Find the Index of the First Occurrence in a String.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        # KMP algorithm 
        
        # first initialize the substring with -1 
        pattern = buildPattern(needle)
        # print(pattern)
        matchingIndex = findMatch(haystack, pattern, needle)
        
        
        return matchingIndex
    
def buildPattern(substring):
    pattern = [-1 for i in range(len(substring))]
    
    i = 0 
    j = 1 
    
    while j < len(substring):
        if substring[i] == substring[j]:
            pattern[j] = i
            i += 1 
            j += 1 
        elif i > 0:
            i = pattern[i - 1] + 1  
        else:
            j += 1 
            
    return pattern

def findMatch(haystack, pattern, needle):
    res = 0
    i = 0 
    j = 0 
    while i < len(haystack) and j < len(needle):
        if haystack[i] == needle[j]:
            i += 1 
            j += 1
            
            if j == len(needle):

                return res 
        elif j > 0:
            j = pattern[j - 1] + 1 
            res = i - j 
        else:
            res = i + 1 
            i += 1 
        # print(i, j)
            
    return res if j == len(needle) else -1
            

    
    return res 

File: strings/28. Find the Index of the First Occurrence in a String.cpp
class Solution {
public:

vector<int> calculateZArray(const string& str) {
    int n = str.length();
    vector<int> z(n, 0);

    int l = 0, r = 0;
    for (int i = 1; i < n; ++i) {
        if (i <= r) {
            z[i] = min(r - i + 1, z[i - l]);
        }
        while (i + z[i] < n && str[z[i]] == str[i + z[i]]) {
            ++z[i];
        }
        if (i + z[i] - 1 > r) {
            l = i;
            r = i + z[i] - 1;
        }
    }

    return z;
}
    int strStr(string needle, string haystack) {
        int m = haystack.length();
        int n = needle.length();
        string total = haystack + "@" + needle;
        vector<int>dp = calculateZArray(total);
        int l = 0, r =0;
        for (int i = 1; i < m + n + 1; i++) {
            if (i > r) {
                l = r = i;
                while (r < m + n + 1 && total[r] == total[r - l]) {
                    r++;
                }
                dp[i] = r - l;
                r--;
            } else {
                int t = i - l;
                if (dp[t] < r - i + 1) {
                    dp[i] = dp[t];
                } else {
                    l = i;
                    while (r < m + n + 1 && total[r] == total[r - l]) {
                        r++;
                    }
                    dp[i] = r - l;
                    r--;
                }
            }
        }
        for (int i = 0; i < m + n + 1; i++) {
            if (dp[i] == haystack.length()) return i - haystack.length() - 1;
        }

        return -1;
    }
};